api/external_auth.proto

```diff
@@ -0,0 +1,151 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The name of the tenant for multi-tenant support.
+  string tenant_name = 1;
+
+  // The request attributes.
+  AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+  // Status OK allows the request. Any other response indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
+    string service = 3;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
+    map<string, string> labels = 4;
+
+    // The validated identity of this peer.
+    // For example, the identity associated with the workload such as a service account.
+    // If an X.509 certificate is used to assert the identity this field should be sourced from
+    // `Subject` and `Subject Alternative Names`.
+    repeated string principals = 5;
+
+    // Other claims about the peer that are not captured in labels and principals.
+    google.protobuf.Struct claims = 6;
+
+    // The CLDR country/region code associated with the above IP address.
+    // If the IP address is private, the `region_code` should reflect the
+    // physical location where this peer is running.
+    string region_code = 7;
+  }
+
+  // This message defines attributes for an HTTP request. If the actual
+  // request is not an HTTP request, the runtime system should try to map
+  // the actual request to an equivalent HTTP request.
+  message Request {
```

Comment:
This should be `HTTPRequest` and we should drop the bit about the runtime system trying to map the actual request into an HTTP request.

I buy the arguments about wanting to keep the description of Peers generic, since that is the job of the authentication layer.  I do *not* think it is the job of the authentication layer to try to canonicalize non-HTTP requests into a generic framework.  This will always be lossy.  Some authZ systems will ask operators to express policy against a single canonical request model, but many will allow operators to express policy against the idioms of the particular application layer protocol. If the former, it should be on the authz system to map the protocols onto the unified model.

Reply:
I don't think we should use HTTP request here. Envoy should map the actual request to the logical request and authz should only apply to logic request. This should simplify authz system quite a bit.

As we discussed in the meeting, the auth system has several layers, protocol translation, authn, authn policy, authz policy. It seems strange for authz system to know how to map REST with gRPC. It means when a new wire protocol is introduced, all existing authz config needs to be updated, which would not be possible in practice.

I think we should limit authz just to "what can you do" and assuming the previous layer already knows the "what" and "you" parts (e.g. the operation and the principal). A monolithic auth system would be too difficult to implement and it is a single point of failure that can not be accepted by large enterprises.

Reply:
I'm not saying that auth systems should not map wire protocols to some canonical request format---but that it should happen on the server side of this API if needed.  Not all authz systems will want to canonicalize the request, and even the ones that do will disagree about the "right" mapping. I think it is folly for us to try to do this in Envoy. The gamut of application layer protocols is incredibly wide. We will end up with compromises no one is happy with.

Instead we should focus on the achievable goal of faithfully representing the request over this API. Adding new protocol support to Envoy will mean extending this API, but it need not mean that authz policy changes. Authz systems will still be free to implement a mapping layer in the way that makes sense for their policy model.

Reply:
http(s), h2 and grpc can all be abstracted correctly here. For `tcp`, `udp` most of the proto is empty. I think this abstraction works well. Spike, are you suggesting that we just change the name of the proto?

Reply:
@spikecurtis We are going to have layers of abstraction in the end. There is no clear right or wrong here. I don't think it makes much sense to let authz system to converge wire requests to logical requests. For example, I don't recall any authz system understands REST, gRPC, SOAP, Thrift, and other hundreds of RPC protocols. It should be API management's job to map incoming request to logical request, such as a principal performs an operation on a resource. The authz system decides whether to allow the request.

If we pass HTTP requests to authz, it means we eventually need to pass another 100 kinds of requests to authz, which would not be a good choice.

Mandar and I have considered many options. Based on our production experience, we believe this proposal has high probability to scale up to hundreds of protocols, millions of services. The abstraction of "logical request" and "multi tenant" allow us to achieve it. We have battle tested the design by running billions of dollars business on our current system.

If we choose to use HTTPRequest, all other protocols will be mapped to HTTPRequest anyway, then the HTTPRequest will become the logical and abstract request, so the end result is the same.

Regarding the operator control, I think the platform designer typically has more insight and global review than operators. Once services interact across teams or companies, the operator decisions will become a huge barrier. Different services can not communicate each other well. For example, if each operator chooses a metric for "requests per second", we will have many thousands or millions of such metric all do the same thing. It would be much better that is a single metric to represent "requests per second between service A and service B", regardless who operate service A or service B.

Reply:
@mandarjog I'm suggesting we change the name of the message, and drop the comment about runtime systems mapping other protocols to HTTP.

@wora While Google may have plenty of experience building large systems, that experience is in the context of building something that works for one company. I'm arguing here for flexibility since many different organizations will be building on these abstractions.

Consider, for example, the challenge of mapping a relational database query into HTTP. These typically involve multiple tables, requiring different privileges on each (insert, select, delete, update). Sure, it **can** be done, in multiple ways. Which way is the correct one? Would any of the ways be satisfactory to a developer whose primary goal is to write an authorization systems for relational databases? No! They want to deal with database abstractions like tables and columns, not headers and URLs.

No one is arguing that service operators write the code that converts the different requests into a canonical model. It should be done by the developers of the authorization systems. If Google wants to map everything to HTTP, go ahead!, but do it on the server side of this interface.

Reply:
I agree with @spikecurtis. I would recommend that we keep this practical and targeted. The proto is built in such a way that it is easy to extend later. This entire structure is very HTTP specific and I'm not sure why we would ask Envoy to map non-HTTP to this. In the future we can easily add `TcpRequest`, `RedisRequest`, etc. I would recommend removing all references to TCP, runtime, etc. and defining a separate `TcpRequest` now if needed, much like we did in the access log. The type of request should be inside a one_of also.